doc/design_docs/strings.md

```diff
@@ -0,0 +1,321 @@
+the very end of its field, with no terminator/padding, and some require the
+terminator/padding.  In general, it seems that terminated strings are more
+likely to require the terminator, while padded strings can usually be entered
+with no padding.
+
+There are, no doubt, other ways of delimiting strings.  These seem to be rare
+and sui generis, and can often be handled by modeling them as length-determined
+strings, then applying the necessary logic in code.
+
+There are also multiple *encodings* for strings, such as ASCII, ISO/IEC 8859-1
+("Latin-1"), UTF-8, UTF-16, etc.  UTF-16 seems to be rare outside of
+Windows-based software.  Hardware almost always appears to use ASCII (encoded as
+one character per byte, with the high bit always clear).
+
+
+## Proposal
+
+### Bytestrings Only
+
+All strings in Emboss should be considered to be opaque blobs of bytes;
+interpretation as ASCII, Latin-1, UTF-8, etc. should be left to the application.
+
+UTF-16 strings are explicitly not handled by this proposal.  In principle, one
+could add a "byte width" parameter to the string types, but it does not seem
+important for now.  This decision can be revisited later.
+
+
+### New Built-In Types
+
+Add three new types to the Prelude (names subject to change):
+
+1.  `FixString`, a string whose contents should be the entire field containing
+    the `FixString`.  When writing to a `FixString`, the value must be exactly
+    the same length as the field.
+
+    `CouldWriteValue()` should return `true` for all strings that are exactly
+    the correct length.
+
+    `FixString` is very close to a notional `Blob` type or the current
+    `UInt:8[]` type, except for differences in text format.
+
+2.  `ZString`, a terminated string.  A `ZString` with no arguments uses a null
+    byte (`'\0'`) as the terminator.  An optional argument can be used to
+    specify the terminator -- a `ZString(36)`, for example, would be terminated
+    by `$`.  When reading, the value returned is all bytes up to, but not
+    including, the first terminator byte.  When writing, for compatibility, the
+    entire field should be written, using the terminator value for padding if
+    there is extra space.  A second optional parameter can be used to specify
+    that the terminator is not required: `ZString(0, false)` can fill the
+    underlying field with no terminator.
+
+    `CouldWriteValue()` should return `true` if the value is no longer than the
+    field and the value does not *contain* any instances of the terminator
+    byte.
+
+3.  `PaddedString`, a padded string.  A `PaddedString` with no arguments uses
+    space (`' '`, 32) as the padding value.  An optional argument can be used to
+    specify the padding -- a `PaddedString(0)`, for example, would be padded
+    with null bytes.  When reading, the end of the string is discovered by
+    walking *backwards* from the end until a non-padding byte is found, then
```

Comment:
It looks like strings are all LTR, should note somewhere that RTL is not yet supported.

Reply:
Unless I'm totally mistaken, LTR and RTL are handled out at the presentation layer -- Emboss does not really support either one, because that level of interpretation happens much later.